Implemented the bulk add feature requested in #223

wforbes5 opened the pull request:
This is my first ever attempt at open source contribution, so apologies in advance if I am not following the proper process.

I did not write any automated tests however I did test by uploading around 20-40 games at a time for n64, ps1, and atari.

Workflow Changes
  1.  Added a bulk upload icon to all of the platform screens. 
  2. Clicking this icon brings user to a new bulk upload endpoint.
  3. Users can drop multiple files into this screen, which will then be matched against the steam grid api, automatically taking the highest quality match.
  4.  Added a file name label to the new endpoint so users can see the file name that was uploaded to evaluate if the match is correct.

Backend
  1. Added an auto-match function to the steamgrid api, this was necessary to simplify the workflow.
  2. Added new function process_new_shortcut so this function could be shared by both the single manual upload functionality and bulk upload process.
  3. Implemented  context manager interface on the ShortcutsFile object that locks the thread and file to prevent bad writes to this file.
  4.  Utilized this interface this in the process_new_shortcut function.

NeroReflex commented:
Thanks! I defer to @alkazar since I can't understand python

alkazar commented:
That is some awesome work. My main concern is overloading the SteamGridDB API limits. We have already had our API token revoked once due to overuse and it is very disruptive.

In order for this to be deployed we would have to create and run some kind of proxy server to SteamGridDB.

wforbes5 commented:
Yea I worried there would be some kind of rate limiting.

Does chimera just have one token that is supplied with the install?

Or is this associated with the users steam account?

The reason I ask is that if we can store that api key when a user logs into steam on the device we can utilize their key to make the calls instead of a shared one.

I’m not sure what the rate limiting level is but I would be surprised if a person even with thousands of roms would be able to be able to hit that limit.

On Fri, Sep 26, 2025 at 12:32 PM Alesh Slovak ***@***.***>
wrote:

> *alkazar* left a comment (ChimeraOS/chimera#326)
> <https://github.com/ChimeraOS/chimera/pull/326#issuecomment-3339509719>
>
> That is some awesome work. My main concern is overloading the SteamGridDB
> API limits. We have already had our API token revoked once due to overuse
> and it is very disruptive.
>
> In order for this to be deployed we would have to create some kind of
> proxy server to SteamGridDB.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ChimeraOS/chimera/pull/326#issuecomment-3339509719>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACZ3JQ33KP7DUXQSTSRWYN33UVTAVAVCNFSM6AAAAACHRRDOOGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTGMZZGUYDSNZRHE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

Commit:
Added the bulk upload functionarlity

wforbes5 commented:
I added a quick change to the change per our discussion  @alkazar. I added in functionality that prevents the bulk upload functionality unless the user has added their own personal API key via the settings section. This should prevent the risk of rate limiting based on the current chimera api key.

NeroReflex commented:
Thanks! I hope alkazar will find it amusing as I do.

wforbes5 commented:
> Thanks! I hope alkazar will find it amusing as I do.

It’s my first ever attempt at contributing to open source. I hope I screw something  up to warrant the humor.
